### ISSUE TITLE
[RFC] [Qwen-Image] WS1 Single-GPU Kernel List

### ISSUE THREAD

**Issue by zhangj1an** [RL-Align/RL-Kernel#386]
## Motivation 

It is time to support multimodal train-infer consistency! This task is open to the community for contribution and I will actively review it. 

## Our 1st model: Qwen-Image

To start with, we will work on Qwen-Image first, it has standard MMDiT structure, with no MoE. Also it is supported by vllm-omni and sglang-omni. Later we can work on bigger models. For your reference, this is the architecture breakdown of Qwen-Image, also a detailed walkthrough of the MMDiT module. 

<img width="2225" height="1932" alt="Image" src="https://github.com/user-attachments/assets/52c56986-13e6-46c1-9207-c3a9f5ca75d7" />

<img width="3462" height="2762" alt="Image" src="https://github.com/user-attachments/assets/c9327cca-d2cb-4203-b219-49f073541945" />

Here is a list of single card ops that needs to be written, open to community for contribution. 

## Kernel table

**Status:** 🙋 open → ⏳ in progress → 👀 in review → ✅ merged

To claim: put your handle in the `GitHub` column and open a PR against this file.

| Kernel | What it does | fwd / bwd | WS2 | GitHub | PR | Status |
|---|---|---|---|---|---|---|
| `latent_pack_unpack` | Patchify / unpatchify permutation and truncation rule | fwd + bwd | yes | @nodeeeeee| | ⏳ |
| `latent_normalize` | `(z − mean) * std_inv` and its inverse, 16 per-channel constants | fwd + bwd | yes | @JLiu4Coding | | ⏳ |
| `latent_proj_gemm` | `img_in` 64→3072 and `proj_out` 3072→64, FP32 throughout | fwd + bwd | no |@haoruilee| | ⏳ |
| `txt_in_rmsnorm_linear` | Text-side RMSNorm(3584) + Linear 3584→3072 | fwd + bwd | no | | | 🙋 |
| `timestep_embed_mlp` | Sinusoidal (256 ch, scale 1000) + 2-layer MLP → 3072 | fwd + bwd | no | | | 🙋 |
| `adaln_modulation` | `chunk(3)` into shift/scale/gate, apply `norm(x)*(1+scale)+shift` | fwd + bwd | yes | @hongleng  | | ⏳ |
| `adaln_gate_residual` | Gated residual write-back `x + gate * sublayer(h)` | fwd + bwd | yes |@BruceLoveDecimal | | ⏳ |
| `qkv_bias_gemm` | 3072 → 3×3072 + bias, one weight set per stream | fwd + bwd | yes (TP col) | | | 🙋 |
| `qk_rmsnorm` | Per-head RMSNorm on Q/K, `head_dim = 128` | fwd + bwd | no | NPU: @erfgss | | ⏳ |
| `multi_axis_rope` | Axis split `[16, 56, 56]`, text on the grid diagonal | fwd + bwd | yes (SP) | NPU: @erfgss | | ⏳ |
| `joint_attn_qk_gemm` | `QKᵀ / √d`, seq = `T_txt + T_img` | fwd + bwd | yes | @ryankert01  | | ⏳ |
| `joint_attn_softmax` | Online softmax over the key sequence, fixed tiling | fwd + bwd | yes | @BPbruce | | ⏳ |
| `joint_attn_av_gemm` | `A · V` | fwd + bwd | yes | | | 🙋 |
| `joint_attention_fused` | Fusion of the three above, must match their composition byte for byte | fwd + bwd | yes (Ulysses / Ring) | | | 🙋 |
| `attn_out_bias_gemm` | Attention output projection 3072→3072 + bias | fwd + bwd | yes (TP row) | | | 🙋 |
| `mlp_up_gemm_gelu` | 3072→12288 + bias + **tanh-approximate** GELU | fwd + bwd | yes (TP col) | | | 🙋 |
| `mlp_down_gemm` | 12288→3072 + bias | fwd + bwd | yes (TP row) | | | 🙋 |
| `flow_sigma_table` | Sigma schedule: dynamic shift + terminal stretch | fwd only | yes (must match across ranks) |@Txy02 | | ⏳ |
| `flow_sde_step_logp` | SDE step `prev_mean` and `logp`, FP32 throughout | fwd + bwd (dlogp) | yes | | | 🙋 |

## How to contribute 

Please refer to #204 as an example to add a new ops. You must include python, triton, cuda implementation, also binding + registry gating + clean fallback + tests + benchmark + validation environment.

1. Fixed reduction tree, frozen and documented. Addition order never changes with batch size, token count, SM count, or any runtime condition.
2. FP32 accumulators everywhere.
3. No Split-K, Stream-K, or atomic partial accumulation.
4. No fast-math, no `__expf` / `__logf`, no TF32, no compiler reassociation.
5. One FP32 → BF16 cast per kernel, at the final output only.
6. **Batch invariance.** A row's output bytes do not depend on who it is batched with or
   how it is tiled. This is the core acceptance test and the precondition for WS2.
7. Trace records: reduction order, accumulator precision, Split-K / Stream-K state, TF32 state, kernel fingerprint, selected backend.

CUDA is the bit-level reference backend. Other backends must match it byte for byte on the same dataset, or register an explicit tolerance profile with the hardware difference written down. No silent fallback.

## Acceptance

Every PR checks off:

- [ ] Forward matches the FP32 CPU reference byte for byte.
- [ ] Backward matches the fixed FP32 reference; multi-path gradient sums use a fixed order.
- [ ] **Batch invariance**: batch size, launch geometry, and tiling do not change a row's bytes.
- [ ] Shapes cover `{1024², 1328², 1664×928}`.
- [ ] Trace records reduction order, accumulator precision, Split-K / Stream-K state, TF32 state, kernel fingerprint.
- [ ] FP32 CPU reference and bit-equality harness delivered.
- [ ] The kernel's own pinned items have explicit test cases.

## WS1 exit criteria

- [ ] All 19 rows ✅.
- [ ] One MMDiT block assembled from WS1 kernels: forward within the tolerance contract,
      `base.grad is None`, LoRA `dA / dB` finite and non-zero.
- [ ] Per-layer error curve across 60 layers produced, for locating first divergence.
- [ ] Single-step `logp` reproducible on fixed inputs and byte-equal to the rollout side.

Then WS2 opens: build `<op>_parallel` for every kernel marked yes, with byte equality against the WS1 version as the acceptance criterion.

## CC list 
if you are interested just ping below!
@Dnoob @Zhifu-Liu @bitborne @Billy1900 @icenfly @JLiu4Coding @haoruilee @erfgss @nodeeeeee @Chen-BUPT @ryankert01 @BruceLoveDecimal @Flink-ddd

**Comment by BruceLoveDecimal** [comment]
hi, can i take **adaln_gate_residual** as a try?

**Comment by Flink-ddd** [comment]
Awesome! I'll recommend this task not just to existing contributors, but also to external developers looking to contribute to our project.

**Comment by JLiu4Coding** [comment]
Hi @zhangj1an , I am so exciited for our new plan. Can I take the latent_normalize?

**Comment by haoruilee** [comment]
Hi @zhangj1an, can I take **latent_proj_gemm**?

**Comment by nodeeeeee** [comment]
Hi @zhangj1an, can I take `latent_pack_unpack`?

**Comment by erfgss** [comment]
Hi @zhangj1an @Flink-ddd ，can I take qk_rmsnorm and multi_axis_rope in Ascend A2/A3,？

**Comment by BPbruce** [comment]
Hi @zhangj1an, I'm interested in this part. Can I take `joint_attn_softmax`?

**Comment by hongleng** [comment]
Hi @zhangj1an @Flink-ddd, I'd like to claim `adaln_modulation` in #386.

I plan one WS1 single-GPU PR with Python, Triton, and CUDA forward/backward implementations, bindings, registry gating, and explicit fallback.

The PR will include fixed-order FP32 reductions, correctness and batch-invariance tests, a bit-equality harness, trace metadata, benchmarks, and the validation environment.

I will start with the standard Qwen-Image path: `x: [B, S, H]` and shared modulation `[B, 3H]` (`H=3072`). The per-block `SiLU + Linear` projection and `adaln_gate_residual` are out of scope. I'll derive the image-token lengths from the VAE/patching code and document the tested GPU, dtypes, and shapes in the PR.

Before implementation, could you clarify two points?

1. Is the edit-specific `modulate_index` path required in the optimized Triton/CUDA kernels for this row, or may it use an explicit PyTorch fallback?
2. #386 asks for byte equality with the FP32 CPU reference and also calls CUDA the bit-level reference. Must CPU and CUDA match byte for byte, or should CPU be an accuracy reference under an agreed tolerance while CUDA is the bit-level reference for other GPU backends?

**Comment by ryankert01** [comment]
I can help with joint_attn_qk_gemm, nice rfc

**Comment by Flink-ddd** [comment]
sure, assigned. Thank you.

**Comment by Txy02** [comment]
Hi， @zhangj1an ，I'm interested in this part. Can I take flow_sigma_table？ Thanks！

**Comment by zzf2024** [comment]
Hi @zhangj1an, I’d like to claim timestep_embed_mlp for the CUDA/A100 path. I plan to implement the Python, Triton, and SM80 CUDA forward/backward paths, including the 256-channel sinusoidal projection with scale 1000 and the 256→3072→3072 SiLU MLP. I’ll include fixed reduction order, FP32 accumulation, explicit fallback, batch-invariance tests, trace metadata, and A100 benchmarks. Before starting, could you clarify the exactness contract for CPU vs CUDA sin/cos and whether guidance/additional conditioning are required in the first PR?
